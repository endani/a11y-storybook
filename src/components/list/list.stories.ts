import type { Meta, StoryObj } from "@storybook/react"
import { fn } from "@storybook/test"
import { List } from "./list"

const mockOptions = [
  "Apple",
  "Banana",
  "Cherry",
  "Date",
  "Elderberry",
  "Fig",
  "Grape",
  "Honeydew",
]

const meta = {
  title: "Example/List",
  component: List,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {},
  args: { onSelect: fn(), options: mockOptions },
} satisfies Meta<typeof List>

export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {
  args: {},
}

export const MaxResults: Story = {
  args: {
    maxResults: 5,
  },
}

export const SearchTerm: Story = {
  args: {
    searchTerm: "App",
  },
}
