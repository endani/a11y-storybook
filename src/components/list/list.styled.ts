import styled from "styled-components"

export const ListWrapper = styled.ul<{ maxResults?: number }>`
  font-family: "Nunito Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  gap: 10px;
  padding: 20px;
  align-items: center;
  width: 100%;
  overflow-y: auto;
  max-height: ${({ maxResults }) =>
    maxResults ? `${maxResults * 50}px` : "auto"};
`

export const Option = styled.li`
  align-items: center;
  border-radius: 8px;
  border: 1px solid #e5e5e5;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  flex-shrink: 0;
  height: 40px;
  padding: 0 16px;
  width: 95%;

  mark {
    background: #99c7fb;
    border-radius: 4px;
    color: #000000;
    font-weight: 600;
    padding: 0 2px;
  }

  &:focus {
    outline: none;
    border: 1px solid #27272a;
  }
`
