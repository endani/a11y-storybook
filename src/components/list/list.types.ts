export interface ListProps {
  options: string[]
  onSelect?: (option: string) => void
  tabIndex?: number
  id?: string
  searchTerm?: string | null
  maxResults?: number
}
