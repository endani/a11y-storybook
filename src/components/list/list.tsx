import { generateMarkedText } from "../../utils"
import { Option, ListWrapper } from "./list.styled"
import { ListProps } from "./list.types"

const List = (props: ListProps) => {
  const { options, id, searchTerm = "", onSelect, maxResults } = props

  const handleOnKeyDown = (event: React.KeyboardEvent) => {
    event.stopPropagation()

    const focusedElement = document.activeElement as HTMLElement
    const focusedElementValue = focusedElement.textContent

    if (event.key === "Enter" && focusedElementValue) {
      onSelect && onSelect(focusedElementValue)
    }
  }

  return (
    <ListWrapper
      id={id}
      data-qa="list"
      role="listbox"
      aria-label="Desert list"
      onKeyDown={handleOnKeyDown}
      maxResults={maxResults}
    >
      {options.map((option, index) => (
        <Option
          aria-posinset={index + 1}
          aria-setsize={options.length}
          tabIndex={0}
          role="option"
          key={index}
          value={option}
          onClick={() => onSelect && onSelect(option)}
          dangerouslySetInnerHTML={{
            __html: generateMarkedText(option, searchTerm as string),
          }}
        />
      ))}
    </ListWrapper>
  )
}

export { List }
