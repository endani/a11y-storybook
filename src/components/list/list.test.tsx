import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import "@testing-library/jest-dom"

import { List } from "./list"
import { ListProps } from "./list.types"

const mockOnSelect = jest.fn()

const renderList = ({ options }: ListProps) =>
  render(<List options={options} onSelect={mockOnSelect} />)

describe("List", () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("should render a List correctly", () => {
    const { container } = renderList({
      options: ["Option 1", "Option 2", "Option 3"],
    })

    expect(container).toMatchSnapshot()
    expect(screen.getByTestId("list")).toBeInTheDocument()
  })

  it("should call onSelect when an option is clicked", async () => {
    renderList({
      options: ["Apple", "Banana", "Orange"],
    })

    const option = screen.getByText("Apple")

    await userEvent.click(option)

    expect(mockOnSelect).toHaveBeenCalled()
    expect(mockOnSelect).toHaveBeenCalledTimes(1)
  })
})
