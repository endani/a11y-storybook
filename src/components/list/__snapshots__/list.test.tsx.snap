// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render a List correctly 1`] = `
<div>
  <ul
    aria-label="Desert list"
    class="sc-aYaIB bDsthc"
    data-qa="list"
    role="listbox"
  >
    <li
      aria-posinset="1"
      aria-setsize="3"
      class="sc-gEvDqW hHYJEe"
      role="option"
      tabindex="0"
      value="Option 1"
    >
      <mark />
      O
      <mark />
      p
      <mark />
      t
      <mark />
      i
      <mark />
      o
      <mark />
      n
      <mark />
       
      <mark />
      1
      <mark />
    </li>
    <li
      aria-posinset="2"
      aria-setsize="3"
      class="sc-gEvDqW hHYJEe"
      role="option"
      tabindex="0"
      value="Option 2"
    >
      <mark />
      O
      <mark />
      p
      <mark />
      t
      <mark />
      i
      <mark />
      o
      <mark />
      n
      <mark />
       
      <mark />
      2
      <mark />
    </li>
    <li
      aria-posinset="3"
      aria-setsize="3"
      class="sc-gEvDqW hHYJEe"
      role="option"
      tabindex="0"
      value="Option 3"
    >
      <mark />
      O
      <mark />
      p
      <mark />
      t
      <mark />
      i
      <mark />
      o
      <mark />
      n
      <mark />
       
      <mark />
      3
      <mark />
    </li>
  </ul>
</div>
`;
